File: svws-core/src/main/java/de/nrw/schule/svws/core/data/kaoa/KAOAEbene4Eintrag.java
package de.nrw.schule.svws.core.data.kaoa;

import de.nrw.schule.svws.core.transpiler.TranspilerDTO;
import de.nrw.schule.svws.core.types.kaoa.KAOAZusatzmerkmal;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotNull;
import jakarta.xml.bind.annotation.XmlRootElement;

/**
 * Objekte dieser Klasse enthalten die im Rahmen von KAoA 
 * gültigen Einträge für die SBO Ebene 4.
 */
@XmlRootElement
@Schema(description="ein Eintrag in dem Katalog der Einträge für die SBO Ebene 4.")
@TranspilerDTO
public class KAOAEbene4Eintrag {

	/** Die ID des Katalog-Eintrags. */
	@Schema(required = true, description = "die ID des Katalog-Eintrags", example="1")
	public long id;

	/** Das Kürzel des Eintrags für die SBO Ebene 4. */
	@Schema(required = true, description = "das Kürzel des Eintrags für die SBO Ebene 4", example="SBO 6.5.1.1")
	public @NotNull String kuerzel = "";
	
	/** Die Beschreibung des Eintrags für die SBO Ebene 4. */
	@Schema(required = true, description = "die Beschreibung des Eintrags für die SBO Ebene 4", example="Langzeitpraktikum 1-tägig")
	public @NotNull String beschreibung = "";

    /** Das Zusatzmerkmal, welcher der Eintrag zugeordnet ist. */
    @Schema(required = true, description = "das Zusatzmerkmal, welcher der Eintrag zugeordnet ist", example="SBO 6.5.1")
    public @NotNull String zusatzmerkmal = "";
	
	/** Gibt an, in welchem Schuljahr der Eintrag einführt wurde. Ist kein Schuljahr bekannt, so ist null gesetzt. */
	@Schema(required = false, description = "gibt an, in welchem Schuljahr der Eintrag einführt wurde. Ist kein Schuljahr bekannt, so ist null gesetzt", example="2020")
	public Integer gueltigVon = null;

	/** Gibt an, bis zu welchem Schuljahr der Eintrag gültig ist. Ist kein Schuljahr bekannt, so ist null gesetzt. */
	@Schema(required = false, description = "gibt an, bis zu welchem Schuljahr der Eintrag gültig ist. Ist kein Schuljahr bekannt, so ist null gesetzt", example="null")
	public Integer gueltigBis = null;


	/**
	 * Erstellt einen KAoA-Eintrag der SBO Ebene 4 mit Standardwerten
	 */
	public KAOAEbene4Eintrag() {
	}

	/**
	 * Erstellt einen KAoA-Eintrag der SBO Ebene 4 mit den angegebenen Werten
	 * 
	 * @param id             die ID
	 * @param kuerzel        das Kürzel 
	 * @param beschreibung   die Beschreibung
	 * @param zusatzmerkmal  das Zusatzmerkmal, dem der Eintrag der SBO Ebene 4 zugeordnet ist
	 * @param gueltigVon     das Schuljahr, wann der Eintrag eingeführt wurde oder null, falls es nicht bekannt ist und "schon immer gültig war"
	 * @param gueltigBis     das Schuljahr, bis zu welchem der Eintrag gültig ist
	 */
	public KAOAEbene4Eintrag(long id, @NotNull String kuerzel, @NotNull String beschreibung, 
	        @NotNull KAOAZusatzmerkmal zusatzmerkmal, Integer gueltigVon, Integer gueltigBis) {
		this.id = id;
		this.kuerzel = kuerzel;
		this.beschreibung = beschreibung;
		this.zusatzmerkmal = zusatzmerkmal.daten.kuerzel;
		this.gueltigVon = gueltigVon;
		this.gueltigBis = gueltigBis;
	}

}

File: svws-webclient/core/src/core/data/kaoa/KAOAEbene4Eintrag.ts
import { JavaObject } from '../../../java/lang/JavaObject';
import { KAOAZusatzmerkmal, cast_de_svws_nrw_core_types_kaoa_KAOAZusatzmerkmal } from '../../../core/types/kaoa/KAOAZusatzmerkmal';

export class KAOAEbene4Eintrag extends JavaObject {

	/**
	 * Die ID des Katalog-Eintrags.
	 */
	public id : number = 0;

	/**
	 * Das Kürzel des Eintrags für die SBO Ebene 4.
	 */
	public kuerzel : string = "";

	/**
	 * Die Beschreibung des Eintrags für die SBO Ebene 4.
	 */
	public beschreibung : string = "";

	/**
	 * Das Zusatzmerkmal, welcher der Eintrag zugeordnet ist.
	 */
	public zusatzmerkmal : string = "";

	/**
	 * Gibt an, in welchem Schuljahr der Eintrag einführt wurde. Ist kein Schuljahr bekannt, so ist null gesetzt.
	 */
	public gueltigVon : number | null = null;

	/**
	 * Gibt an, bis zu welchem Schuljahr der Eintrag gültig ist. Ist kein Schuljahr bekannt, so ist null gesetzt.
	 */
	public gueltigBis : number | null = null;


	/**
	 * Erstellt einen KAoA-Eintrag der SBO Ebene 4 mit Standardwerten
	 */
	public constructor();

	/**
	 * Erstellt einen KAoA-Eintrag der SBO Ebene 4 mit den angegebenen Werten
	 *
	 * @param id             die ID
	 * @param kuerzel        das Kürzel
	 * @param beschreibung   die Beschreibung
	 * @param zusatzmerkmal  das Zusatzmerkmal, dem der Eintrag der SBO Ebene 4 zugeordnet ist
	 * @param gueltigVon     das Schuljahr, wann der Eintrag eingeführt wurde oder null, falls es nicht bekannt ist und "schon immer gültig war"
	 * @param gueltigBis     das Schuljahr, bis zu welchem der Eintrag gültig ist
	 */
	public constructor(id : number, kuerzel : string, beschreibung : string, zusatzmerkmal : KAOAZusatzmerkmal, gueltigVon : number | null, gueltigBis : number | null);

	/**
	 * Implementation for method overloads of 'constructor'
	 */
	public constructor(__param0? : number, __param1? : string, __param2? : string, __param3? : KAOAZusatzmerkmal, __param4? : null | number, __param5? : null | number) {
		super();
		if ((typeof __param0 === "undefined") && (typeof __param1 === "undefined") && (typeof __param2 === "undefined") && (typeof __param3 === "undefined") && (typeof __param4 === "undefined") && (typeof __param5 === "undefined")) {
			// empty method body
		} else if (((typeof __param0 !== "undefined") && typeof __param0 === "number") && ((typeof __param1 !== "undefined") && (typeof __param1 === "string")) && ((typeof __param2 !== "undefined") && (typeof __param2 === "string")) && ((typeof __param3 !== "undefined") && ((__param3 instanceof JavaObject) && ((__param3 as JavaObject).isTranspiledInstanceOf('de.svws_nrw.core.types.kaoa.KAOAZusatzmerkmal')))) && ((typeof __param4 !== "undefined") && (typeof __param4 === "number") || (__param4 === null)) && ((typeof __param5 !== "undefined") && (typeof __param5 === "number") || (__param5 === null))) {
			const id : number = __param0 as number;
			const kuerzel : string = __param1;
			const beschreibung : string = __param2;
			const zusatzmerkmal : KAOAZusatzmerkmal = cast_de_svws_nrw_core_types_kaoa_KAOAZusatzmerkmal(__param3);
			const gueltigVon : number | null = __param4;
			const gueltigBis : number | null = __param5;
			this.id = id;
			this.kuerzel = kuerzel;
			this.beschreibung = beschreibung;
			this.zusatzmerkmal = zusatzmerkmal.daten.kuerzel;
			this.gueltigVon = gueltigVon;
			this.gueltigBis = gueltigBis;
		} else throw new Error('invalid method overload');
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.svws_nrw.core.data.kaoa.KAOAEbene4Eintrag'].includes(name);
	}

	public static transpilerFromJSON(json : string): KAOAEbene4Eintrag {
		const obj = JSON.parse(json);
		const result = new KAOAEbene4Eintrag();
		if (typeof obj.id === "undefined")
			 throw new Error('invalid json format, missing attribute id');
		result.id = obj.id;
		if (typeof obj.kuerzel === "undefined")
			 throw new Error('invalid json format, missing attribute kuerzel');
		result.kuerzel = obj.kuerzel;
		if (typeof obj.beschreibung === "undefined")
			 throw new Error('invalid json format, missing attribute beschreibung');
		result.beschreibung = obj.beschreibung;
		if (typeof obj.zusatzmerkmal === "undefined")
			 throw new Error('invalid json format, missing attribute zusatzmerkmal');
		result.zusatzmerkmal = obj.zusatzmerkmal;
		result.gueltigVon = typeof obj.gueltigVon === "undefined" ? null : obj.gueltigVon === null ? null : obj.gueltigVon;
		result.gueltigBis = typeof obj.gueltigBis === "undefined" ? null : obj.gueltigBis === null ? null : obj.gueltigBis;
		return result;
	}

	public static transpilerToJSON(obj : KAOAEbene4Eintrag) : string {
		let result = '{';
		result += '"id" : ' + obj.id + ',';
		result += '"kuerzel" : ' + JSON.stringify(obj.kuerzel!) + ',';
		result += '"beschreibung" : ' + JSON.stringify(obj.beschreibung!) + ',';
		result += '"zusatzmerkmal" : ' + JSON.stringify(obj.zusatzmerkmal!) + ',';
		result += '"gueltigVon" : ' + ((!obj.gueltigVon) ? 'null' : obj.gueltigVon) + ',';
		result += '"gueltigBis" : ' + ((!obj.gueltigBis) ? 'null' : obj.gueltigBis) + ',';
		result = result.slice(0, -1);
		result += '}';
		return result;
	}

	public static transpilerToJSONPatch(obj : Partial<KAOAEbene4Eintrag>) : string {
		let result = '{';
		if (typeof obj.id !== "undefined") {
			result += '"id" : ' + obj.id + ',';
		}
		if (typeof obj.kuerzel !== "undefined") {
			result += '"kuerzel" : ' + JSON.stringify(obj.kuerzel!) + ',';
		}
		if (typeof obj.beschreibung !== "undefined") {
			result += '"beschreibung" : ' + JSON.stringify(obj.beschreibung!) + ',';
		}
		if (typeof obj.zusatzmerkmal !== "undefined") {
			result += '"zusatzmerkmal" : ' + JSON.stringify(obj.zusatzmerkmal!) + ',';
		}
		if (typeof obj.gueltigVon !== "undefined") {
			result += '"gueltigVon" : ' + ((!obj.gueltigVon) ? 'null' : obj.gueltigVon) + ',';
		}
		if (typeof obj.gueltigBis !== "undefined") {
			result += '"gueltigBis" : ' + ((!obj.gueltigBis) ? 'null' : obj.gueltigBis) + ',';
		}
		result = result.slice(0, -1);
		result += '}';
		return result;
	}

}

export function cast_de_svws_nrw_core_data_kaoa_KAOAEbene4Eintrag(obj : unknown) : KAOAEbene4Eintrag {
	return obj as KAOAEbene4Eintrag;
}
